lib/datadog/tracing.rb

```diff
@@ -0,0 +1,198 @@
+module Datadog
+  # Datadog APM tracing public API.
+  #
+  # The Datadog teams ensures that public methods in this module
+  # only receive backwards compatible changes, and breaking changes
+  # will only occur in new major versions releases.
+  # @public_api
+  module Tracing
+    class << self
+      # (see Datadog::Tracer#trace)
+      # @public_api
+      def trace(name, continue_from: nil, **span_options, &block)
+        tracer.trace(name, continue_from: continue_from, **span_options, &block)
+      end
+
+      # (see Datadog::Tracer#continue_trace!)
+      # @public_api
+      def continue_trace!(digest, &block)
+        tracer.continue_trace!(digest, &block)
+      end
+
+      # The currently active {Datadog::Tracer} instance.
+      #
+      # The instance returned can change throughout the lifetime of the application.
+      # This means it is not advisable to cache it.
+      #
+      # The trace can be configured through {.configure},
+      # through {Datadog::Configuration::Settings::DSL::Tracer} options.
+      #
+      # TODO: This next paragraph can be better written.
+      #
+      # Most of the functionality available through the {.tracer} instance is
+      # also available in public methods in the {Datadog::Tracing} module.
+      # It is preferable to use the public methods in the {Datadog::Tracing} when possible
+      # as {Datadog::Tracing} strongly defines the tracing public API, and thus
+      # we strive to no introduce breaking changes to {Datadog::Tracing} methods.
```

Comment:
Coming back to this: I'm looking at the list of methods in the tracer

```ruby
(Datadog.tracer.public_methods - Object.public_methods).sort
=> 
[:active_correlation, # ✅ in Datadog::Tracing
 :active_span, # ✅ in Datadog::Tracing
 :active_trace, # ✅ in Datadog::Tracing
 :continue_trace!, # ✅ in Datadog::Tracing
 :default_service,
 :default_service=,
 :enabled,
 :enabled=,
 :provider,
 :sampler,
 :set_tags,
 :shutdown!, # ✅ in Datadog::Tracing
 :tags,
 :trace, # ✅ in Datadog::Tracing
 :trace_completed,
 :trace_flush,
 :writer,
 :writer=]
```

and I'm wondering -- what are the remaining usecases not covered by `Datadog::Tracing`?

Could we perhaps add them here, and then make `Datadog::Tracing.tracer` and `Datadog.tracer` private?

Reply:
I'd exposed `enabled` as well now.
For the remaining methods, I don't see the need for them to be exposed today.

BUT, I don't think we should hide the `tracer` instance today, because it has always been advertised as the main point of interaction with `ddtrace`. Even though I don't think we should directly expose, for example, `Datadog.tracer.writer`, I don't have the confidence that no one has a valid use case for this access.

I've tagged `#tracer` here as deprecated.

Reply:
> I'd exposed `enabled` as well now.

Is it by design that we expose the writer for `enabled` as well? That seems weird... Is it just me or doing `Datadog.tracer.enabled = false` while the tracer is active will cause all sorts of havoc?

> BUT, I don't think we should hide the tracer instance today, because it has always been advertised as the main point of interaction with ddtrace.

Hmm I'm not entirely convinced, but my role in the tracing bits is just as an advisor, so I'll trust y'all on this one :)

Reply:
Yeah, I'd like to hide the `tracer` object as well, if at all possible.

> Also, I don't quite understand our overall intention here. We're exposing the `Tracer` object, and have it marked as a public api, so we're never going to change that API either, so... what's the advantage of using these methods again?

I think the `Tracer` object should be hidden such that the only means of doing trace operations would be through `Datadog::Tracing#some_public_method`. By hiding this, we could then further modify the behavior of the `Tracer` in 1.x while maintaining the public API (`Datadog::Tracing`) to avoid breaking changes.

Reply:
Looking at this more, we have to migrate existing integrations to this new public API. Once we do this, and its clear we don't need to expose this for any good reason, we can remove this from the public API.

So consider this as something we will do in the next PR.

Reply:
Just to clarify, from my understanding, what what we're proposing is:

1. Removing the `@public_api` at the top of `class Tracer` that was added in a previous 1.0 branch PR: 

https://github.com/DataDog/dd-trace-rb/blob/e6618175702aa3d2064dd44abe865596b18a9883/lib/ddtrace/tracer.rb#L27-L28

2. Making `Datadog::Tracing.tracer` private in `tracing.rb`.

If my understanding is correct, I believe both changes could already be made in this PR without changing any of the integrations. Especially 2. is a bit weird to plan to change later, since we're adding it in this PR.